gpkit/tools.py

```diff
@@ -112,3 +115,53 @@ def mdmake(filename, make_tex=True):
         with open(filename+".tex.md", "w") as f:
             f.write(texmd)
     return open(filename+".py")
+
+
+def bound_all_variables(model, eps=1e-30, lower=None, upper=None):
+    "Returns model with additional constraints bounding all free variables"
+    lb = lower if lower else eps
+    ub = upper if upper else 1/eps
+    constraints = []
+    for varkey in model.varkeys:
+        units = varkey.descr.get("units", 1)
+        constraints.append([ub*units >= Variable(**varkey.descr),
+                            Variable(**varkey.descr) >= lb*units])
+    m = Model(model.cost, [constraints, model], model.substitutions)
+    m.bound_all = {"lb": lb, "ub": ub, "varkeys": model.varkeys}
+    return m
+
+
+# pylint: disable=too-many-locals
+def determine_unbounded_variables(model, solver=None, verbosity=0,
+                                  eps=1e-30, lower=None, upper=None, **kwargs):
+    "Returns labeled dictionary of unbounded variables."
+    m = bound_all_variables(model, eps, lower, upper)
+    sol = m.solve(solver, verbosity, **kwargs)
+    lam = sol["sensitivities"]["la"][1:]
+    out = defaultdict(list)
+    for i, varkey in enumerate(m.bound_all["varkeys"]):
+        lam_gt, lam_lt = lam[2*i], lam[2*i+1]
+        if lam_gt and lam_lt:  # to avoid divide-by-zero
+            sens_ratio = lam_gt/lam_lt
+        elif lam_gt:
+            sens_ratio = np.inf
+        else:
+            sens_ratio = 0.0
+        if sens_ratio >= 2:  # arbitrary threshold
+            out["pushing up"].append(varkey)
+        elif sens_ratio <= 0.5:  # arbitrary threshold
+            out["pushing down"].append(varkey)
+        else:
+            value = mag(sol["variables"][varkey])
+            distance_below = np.log(value/m.bound_all["lb"])
+            distance_above = np.log(m.bound_all["ub"]/value)
+            if distance_below <= 3:
+                out["hit eps lower bound"].append(varkey)
+            elif distance_above <= 3:
+                out["hit 1/eps upper bound"].append(varkey)
+           #  if distance_below <= 3 or distance_above <= 3:
+           #      # if we're within ~10x of a boundary (arbitrary threshold)
+           #      raise AttributeError("%s appears to have hit a boundary"
+           #                           " but it's not showing up in the"
+           #                           " sensitivities." % varkey)
```

Comment:
The problem with position is that sometimes the product of two numbers is
what's upper unbounded, and so they might not hit the boundary
individually, but are still sensitive to the upper bound.
On Mar 25, 2016 12:50, "Warren Hoburg" <notifications@github.com> wrote:

> In gpkit/tools.py
> <https://github.com/hoburg/gpkit/pull/571#discussion_r57462679>:
>
> > +            out["pushing up"].append(varkey)
> > +        elif sens_ratio <= 0.5:  # arbitrary threshold
> > +            out["pushing down"].append(varkey)
> > +        else:
> > +            value = mag(sol["variables"][varkey])
> > +            distance_below = np.log(value/m.bound_all["lb"])
> > +            distance_above = np.log(m.bound_all["ub"]/value)
> > +            if distance_below <= 3:
> > +                out["hit eps lower bound"].append(varkey)
> > +            elif distance_above <= 3:
> > +                out["hit 1/eps upper bound"].append(varkey)
> > +           #  if distance_below <= 3 or distance_above <= 3:
> > +           #      # if we're within ~10x of a boundary (arbitrary threshold)
> > +           #      raise AttributeError("%s appears to have hit a boundary"
> > +           #                           " but it's not showing up in the"
> > +           #                           " sensitivities." % varkey)
>
> The sensitivity to the bounds in your unit test example is 1.00697474e-06
> - and this decreases as the exponent 1.0000001 is moved closer to 1.
>
> So, the necessary absolute threshold on sensitivity is in some sense a
> measure of how dual infeasible the model is -- how close that exponent is
> to making the model dual feasible.
>
> I think as a practical matter setting a bound of 1e30 and looking for
> absolute position might be most reliable. It's also perhaps more intuitive
> for users to hear that a variable in their model is optimally near 1e30
> -- that might be more understandable than hearing it exceeded the
> sensitivity threshold of 1e-6, even if the latter is a stronger condition.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/hoburg/gpkit/pull/571/files/df5a416930d67c14369ba442ac71ae614255fc53#r57462679>
>